File: package/go/src/count_connections.go
package src

// Log represents a connection log entry at a specific time
type Log struct {
	Time        int // Time 0 <= time <= 1000
	NConnect    int // Number of new connections at this time
	NDisconnect int // Number of new disconnections at this time
}

// CountConnectionsParam represents the parameters for counting connections
type CountConnectionsParam struct {
	EndTime int   // End time for aggregation
	Period  int   // Aggregation interval
	Logs    []Log // Connection logs
}

// CountConnections calculates the number of connections for each period
func CountConnections(param CountConnectionsParam) []int {
	// Initialize arrays to track connections and disconnections
	connect := make([]int, param.EndTime+1)
	disconnect := make([]int, param.EndTime+1)

	// Populate arrays with log data
	for _, log := range param.Logs {
		if log.Time <= param.EndTime {
			connect[log.Time] = log.NConnect
			disconnect[log.Time] = log.NDisconnect
		}
	}

	// Calculate connection count for each period
	var result []int
	for t := 0; t <= param.EndTime; t += param.Period {
		// Sum connections and disconnections up to time t
		totalConnect := 0
		totalDisconnect := 0

		for i := 0; i <= t; i++ {
			totalConnect += connect[i]
			totalDisconnect += disconnect[i]
		}

		// Current connection count = total connections - total disconnections
		currentConnections := totalConnect - totalDisconnect
		result = append(result, currentConnections)
	}

	return result
}

File: package/typescript/src/countConnections.ts
interface Log {
  time: number; // 時刻 0 <= time <= 1000
  nConnect: number; // この時間に新規接続した端末の台数
  nDisconnect: number; // この時間に新規切断した端末の台数
}

interface CountConnectionsParam {
  endTime: number; // 集計を終了する時刻
  period: number; // 集計する間隔
  logs: Log[]; // ログ
}

/**
 * ある期間ごとの接続数を出力する関数.
 *
 * @param param - 接続情報
 * @returns ある期間ごとの接続数
 */
export function countConnections(param: CountConnectionsParam): number[] {
  const connect = new Array(param.endTime + 1).fill(0);
  const disconnect = new Array(param.endTime + 1).fill(0);

  for (const log of param.logs) {
    connect[log.time] = log.nConnect;
    disconnect[log.time] = log.nDisconnect;
  }

  const result: number[] = [];
  for (let t = 0; t <= param.endTime; t += param.period) {
    const totalConnect = connect.slice(0, t + 1).reduce((sum, val) => sum + val, 0);
    const totalDisconnect = disconnect.slice(0, t + 1).reduce((sum, val) => sum + val, 0);
    result.push(totalConnect - totalDisconnect);
  }

  return result;
}

export type { Log, CountConnectionsParam };
